pkg/optimize/sg/reduceCubes.go

```diff
@@ -0,0 +1,142 @@
+package sgoptimizer
+
+import (
+	"slices"
+
+	"github.com/np-guard/vpc-network-config-synthesis/pkg/optimize"
+)
+
+func reduceCubesWithSGRemote(cubes *sgCubesPerProtocol) {
+	deleteOtherProtocolIfAnyProtocolExistsSGRemote(cubes)
+	compressThreeProtocolsToAnyProtocolSGRemote(cubes)
+}
+
+// delete other protocols rules if any protocol rule exists
+func deleteOtherProtocolIfAnyProtocolExistsSGRemote(cubes *sgCubesPerProtocol) {
+	for _, sgName := range cubes.anyProtocol {
+		delete(cubes.tcp, sgName)
+		delete(cubes.udp, sgName)
+		delete(cubes.icmp, sgName)
+	}
+}
+
+// merge tcp, udp and icmp rules into any protocol rule
+func compressThreeProtocolsToAnyProtocolSGRemote(cubes *sgCubesPerProtocol) {
+	for sgName, tcpPorts := range cubes.tcp {
+		if udpPorts, ok := cubes.udp[sgName]; ok {
+			if ic, ok := cubes.icmp[sgName]; ok {
+				if ic.IsAll() && optimize.IsAllPorts(tcpPorts) && optimize.IsAllPorts(udpPorts) {
+					delete(cubes.tcp, sgName)
+					delete(cubes.udp, sgName)
+					delete(cubes.icmp, sgName)
+					cubes.anyProtocol = append(cubes.anyProtocol, sgName)
+				}
+			}
+		}
+	}
+}
+
+// observation: It pays to switch to all protocol rule when we have rules that cover all other protocols
+// on exactly the same cidr (only one protocol can exceed).
+//
+//nolint:gocyclo // multiple checks
+func reduceIPCubes(cubes *ipCubesPerProtocol) {
+	tcpPtr := 0
+	udpPtr := 0
+	icmpPtr := 0
+
+	for tcpPtr < len(cubes.tcp) && udpPtr < len(cubes.udp) && icmpPtr < len(cubes.icmp) {
+		if !optimize.IsAllPorts(cubes.tcp[tcpPtr].Right) { // not all tcp ports
+			tcpPtr++
+			continue
+		}
+		if !optimize.IsAllPorts(cubes.udp[udpPtr].Right) { // not all udp ports
+			udpPtr++
+			continue
+		}
+		if !cubes.icmp[icmpPtr].Right.IsAll() { // not all icmp types & codes
+			icmpPtr++
+			continue
+		}
+
+		// all three protocols include all ports and types & codes
+		// attempt to convert to any protocol rule
+		if compressedToAnyProtocolCubeIPCubes(cubes, tcpPtr, udpPtr, icmpPtr) { // converted to any protocol rule
+			continue
+		}
+
+		// could not compress to any protocol rule -- advance one ipblock
+		// case 1: one protocol ipb contains two other ipbs ==> advance the smaller one
+		// case 2: advance the smaller ipb
+		tcpIPblock := cubes.tcp[tcpPtr].Left
+		udpIPblock := cubes.udp[udpPtr].Left
+		icmpIPblock := cubes.icmp[icmpPtr].Left
+
+		udpComparedToICMP := udpIPblock.Compare(icmpIPblock)
+		tcpComparedToUDP := tcpIPblock.Compare(udpIPblock)
+		tcpComparedToICMP := tcpIPblock.Compare(icmpIPblock)
+
+		udpAndICMPSubsetOfTCP := udpIPblock.IsSubset(tcpIPblock) && icmpIPblock.IsSubset(tcpIPblock)
+		tcpAndICMPSubsetOfUDP := icmpIPblock.IsSubset(udpIPblock) && tcpIPblock.IsSubset(udpIPblock)
+		tcpAndUDPSubsetOfICMP := tcpIPblock.IsSubset(icmpIPblock) && udpIPblock.IsSubset(icmpIPblock)
+
+		switch {
+		// case 1
+		case udpAndICMPSubsetOfTCP && udpComparedToICMP == -1:
+			udpPtr++
+		case udpAndICMPSubsetOfTCP && udpComparedToICMP == 1:
+			icmpPtr++
```

Comment:
Will it be wise to advance `udpPtr` even if `udpComparedToICMP == 0`

Reply:
`udpComparedToICMP == 0` means `udpIP.Equal(IcmpIP)` --> both IP cubes are replaced to anyProtocol cube in `compressedToAnyProtocolCubeIPCubes` func

Reply:
true, but now you rely on the good memory of whoever reads the code.
How about replacing the second case to just
```
case udpAndICMPSubsetOfTCP:
```
?